### PR TITLE
 [4.7.0] multichain

### PR DESCRIPTION
**Pull request by DZariusz**
### Changed
- use `timestamp` as block ID
- new multichain architecture
- address Peckshield audit suggestions

### CODE REVIEW

File: contracts/Chain.sol [hunk 1]
```diff
@@ -126,23 +45,19 @@ contract Chain is BaseChain {
         uint8[] memory _v,
         bytes32[] memory _r,
         bytes32[] memory _s
-    ) public { // it could be external, but for external we got stack too deep
-        uint32 lastBlockId = getLatestBlockId();
-        uint32 lastDataTimestamp = squashedRoots[lastBlockId].extractTimestamp();
+    ) external {
+        // below two checks are only for pretty errors, so we can safe gas and allow for raw revert
```
Comment (avoloshko) [inline contracts/Chain.sol:49]: I wonder if we should  remove all the commented code from contracts

File: contracts/Chain.sol [hunk 2]
```diff
@@ -126,23 +45,19 @@ contract Chain is BaseChain {
...
-            if (lastDataTimestamp + padding >= block.timestamp) revert BlockSubmittedToFast();
-        }
+        _verifySubmitTimestampAndIncSequence(_dataTimestamp);
 
-        if (_dataTimestamp <= lastDataTimestamp) revert DataToOld();
         // we can't expect minter will have exactly the same timestamp
         // but for sure we can demand not to be off by a lot, that's why +3sec
         // temporary remove this condition, because recently on ropsten we see cases when minter/node
         // can be even 100sec behind
         // require(_dataTimestamp <= block.timestamp + 3,
```
Comment (avoloshko) [inline contracts/Chain.sol:59]: I wonder if we should  remove all the commented code from contracts

**Review by avoloshko — APPROVED** [review]

File: contracts/Chain.sol (hunk 1 above)
Comment (DZariusz) [reply contracts/Chain.sol:49]: we agree to remove in next release

File: contracts/Chain.sol (hunk 2 above)
Comment (DZariusz) [reply contracts/Chain.sol:59]: we agree to remove in next release

**Pull request merged**